File: server/protocol.go
package main

import "encoding/json"

type Card string

const (
	MsgJoin          = "join"
	MsgVote          = "vote"
	MsgReveal        = "reveal"
	MsgReset         = "reset"
	MsgRevote        = "revote"
	MsgSetStory      = "setStory"
	MsgSetAutoReveal = "setAutoReveal"
	MsgSetSpectator  = "setSpectator"
	MsgNudge         = "nudge" // both directions: in {targetId}, out {targetId, fromId}

	MsgState  = "state"
	MsgJoined = "joined"
	MsgError  = "error"
)

type Inbound struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

type Outbound struct {
	Type    string `json:"type"`
	Payload any    `json:"payload,omitempty"`
}

type JoinPayload struct {
	RoomID    string `json:"roomId"`
	Name      string `json:"name"`
	RejoinID  string `json:"rejoinId,omitempty"`
	Deck      []Card `json:"deck,omitempty"`
	Spectator bool   `json:"spectator,omitempty"`
	Create    bool   `json:"create,omitempty"`
}

type VotePayload struct {
	Card Card `json:"card"`
}

type SetStoryPayload struct {
	Story string `json:"story"`
}

type SetAutoRevealPayload struct {
	Enabled bool `json:"enabled"`
}

type SetSpectatorPayload struct {
	Spectator bool `json:"spectator"`
}

// NudgePayload is the inbound nudge request: who to buzz.
type NudgePayload struct {
	TargetID string `json:"targetId"`
}

// NudgeEvent is the outbound ephemeral nudge, broadcast to the whole room so
// the target shakes + buzzes and everyone else sees the target's card shake.
// Names are resolved client-side from the participant list, so only IDs travel.
type NudgeEvent struct {
	TargetID string `json:"targetId"`
	FromID   string `json:"fromId"`
}

type StatePayload struct {
	RoomID       string              `json:"roomId"`
	Revealed     bool                `json:"revealed"`
	RevealAt     int64               `json:"revealAt"`  // unix ms; 0 unless countdown in progress
	ServerNow    int64               `json:"serverNow"` // unix ms at send; lets clients correct clock skew for the countdown
	AutoReveal   bool                `json:"autoReveal"`
	Story        string              `json:"story"`
	Deck         []Card              `json:"deck"`
	Participants []ParticipantPublic `json:"participants"`
	History      []HistoryEntry      `json:"history"`
}

type HistoryEntry struct {
	Story string        `json:"story"`
	Votes []HistoryVote `json:"votes"`
	At    int64         `json:"at"`
}

type HistoryVote struct {
	Name string `json:"name"`
	Vote Card   `json:"vote"`
}

type ParticipantPublic struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	HasVoted  bool   `json:"hasVoted"`
	Connected bool   `json:"connected"`
	Spectator bool   `json:"spectator"`
	Vote      *Card  `json:"vote"`
}

type JoinedPayload struct {
	ParticipantID string `json:"participantId"`
}

type ErrorPayload struct {
	Message string `json:"message"`
}

File: web/src/protocol.ts
export type Card = string

export interface ParticipantPublic {
  id: string
  name: string
  hasVoted: boolean
  connected: boolean
  spectator: boolean
  vote: Card | null
}

export interface StatePayload {
  roomId: string
  revealed: boolean
  revealAt: number
  autoReveal: boolean
  story: string
  deck: Card[]
  participants: ParticipantPublic[]
  history: HistoryEntry[]
}

export interface HistoryEntry {
  story: string
  votes: HistoryVote[]
  at: number
}

export interface HistoryVote {
  name: string
  vote: Card
}

export interface JoinedPayload {
  participantId: string
}

export interface ErrorPayload {
  message: string
}

export type InboundFromServer =
  | { type: 'state'; payload: StatePayload }
  | { type: 'joined'; payload: JoinedPayload }
  | { type: 'error'; payload: ErrorPayload }

export type OutboundToServer =
  | {
      type: 'join'
      payload: {
        roomId: string
        name: string
        rejoinId?: string
        deck?: Card[]
        spectator?: boolean
        create?: boolean
      }
    }
  | { type: 'vote'; payload: { card: Card } }
  | { type: 'reveal' }
  | { type: 'reset' }
  | { type: 'revote' }
  | { type: 'setStory'; payload: { story: string } }
  | { type: 'setAutoReveal'; payload: { enabled: boolean } }
